File: src/linked_lists/java/Palindrome.java
package linked_lists.java;

import static linked_lists.java.Node.printLinkedList;

import java.util.Arrays;

/**
 *
 */
public class Palindrome {

  public static void main(String[] args) {
    test(new int[]{1, 2, 3, 4, 3, 2, 1});
    test(new int[]{1});
    test(new int[]{1, 2, 2, 1});
    test(new int[]{1, 2, 3, 4});
  }

  private static void test(int[] arr) {
    char[] chars = new char[20];
    Arrays.fill(chars, '-');
    String delimiter = String.valueOf(chars);
    System.out.println(delimiter);
    System.out.println();

    System.out.println("Linked list:");
    Node list = arrayToLinkedList(arr);
    printLinkedList(list);

    System.out.print("Result of normal method:");
    System.out.println(isPalindrome(list) ? "Palindrome" : "Not palindrome");

    System.out.print("Result of recursive method:");
    System.out.println(isPalindromeRecursive(list) ? "Palindrome" : "Not palindrome");
    System.out.println();
  }

  private static boolean isPalindrome(Node head) {
    Node rev = reversedList(head);
    return isEqual(head, rev);
  }

  private static Node reversedList(Node head) {
    Node re = null;
    for (Node ptr = head; ptr != null; ptr = ptr.next) {
      Node temp = new Node();
      temp.data = ptr.data;
      temp.next = re;
      re = temp;
    }

    return re;
  }

  /*
   * Assume that two lists do NOT have circles.
   */
  private static boolean isEqual(Node head1, Node head2) {
    while (head1 != null && head2 != null) {
      if (head1.data != head2.data) {
        return false;
      }
      head1 = head1.next;
      head2 = head2.next;
    }

    return head1 == null && head2 == null;
  }

  static Node arrayToLinkedList(int[] arr) {
    if (arr.length == 0) {
      return null;
    }

    Node head = null;
    for (int i = arr.length - 1; i >= 0; i--) {
      Node temp = new Node();
      temp.data = arr[i];
      temp.next = head;
      head = temp;
    }

    return head;
  }

  private static boolean isPalindromeRecursive(Node head) {
    int len = getLength(head);
    return isPalindromeRecursive(head, len).isPalindrome;
  }

  static int getLength(Node head) {
    Node ptr = head;
    int  len = 0;
    while (ptr != null) {
      len++;
      ptr = ptr.next;
    }
    return len;
  }

  private static Result isPalindromeRecursive(Node head, int len) {
    if (len == 1) {
      return new Result(head, true);
    }
    if (len == 2) {
      return new Result(head.next, true);
    }

    Result subListResult = isPalindromeRecursive(head.next, len - 2);

    Node tail = subListResult.node.next;
    if (!subListResult.isPalindrome) {
      return new Result(tail, false);
    }

    return new Result(tail, tail.data == head.data);
  }

  /*
   * Store recursive method result, especially for node to store the tail node of sublist.
   */
  private static class Result {

    Node    node;
    boolean isPalindrome;

    Result(Node node, boolean isPalindrome) {
      this.node = node;
      this.isPalindrome = isPalindrome;
    }
  }
}

File: src/linked_lists/kotlin/Palindrome.kt
package linked_lists.kotlin

/**
 *
 */
fun main(args: Array<String>) {
    testPalindrome(intArrayOf(1, 2, 3, 4, 3, 2, 1))
    testPalindrome(intArrayOf(1))
    testPalindrome(intArrayOf(1, 2, 3, 3, 2, 1))
    testPalindrome(intArrayOf(1, 2, 3, 4))
}

private fun testPalindrome(arr: IntArray) {
    val delimiter = String(CharArray(20, { '-' }))
    println(delimiter)
    println()

    println("Linked list:")
    val list = arrayToLinkedList(arr)
    list?.let {
        printLinkedList(it)

        print("Result of normal method:")
        println(if (isPalindrome(list)) "Palindrome" else "Not palindrome")

        print("Result of recursive method:")
        println(if (isPalindromeRecursive(list)) "Palindrome" else "Not palindrome")
    }
    println()
}

private class Result(val node: Node, val isPalindrome: Boolean)

fun isPalindromeRecursive(list: Node): Boolean {
    val length = getLength(list)
    return isPalindromeRecur(list, length).isPalindrome
}

fun getLength(list: Node): Int {
    var length = 0
    var ptr = list as? Node
    while (ptr != null) {
        length++
        ptr = ptr.next
    }
    return length
}

private fun isPalindromeRecur(list: Node, length: Int): Result {
    if (length == 1) {
        return Result(list, true)
    }
    if (length == 2) {
        return Result(list.next!!, true)
    }

    val subListResult = isPalindromeRecur(list.next!!, length - 2)
    val tail = subListResult.node.next!!
    if (!subListResult.isPalindrome) return Result(tail, false)
    return Result(tail, tail.data == list.data)
}

fun isPalindrome(head: Node): Boolean {
    val rev = reversedList(head)
    return isEqual(head, rev)
}

fun reversedList(head: Node): Node? {
    var re: Node? = null
    var ptr = head as? Node
    while (ptr != null) {
        val temp = Node()
        temp.data = ptr.data
        temp.next = re
        re = temp
        ptr = ptr.next
    }
    return re
}

fun isEqual(head1: Node?, head2: Node?): Boolean {
    var ptr1 = head1
    var ptr2 = head2
    while (ptr1 != null && ptr2 != null) {
        if (ptr1.data != ptr2.data) return false

        ptr1 = ptr1.next
        ptr2 = ptr2.next
    }

    return ptr1 == null && ptr2 == null
}

fun arrayToLinkedList(arr: IntArray): Node? {
    var head: Node? = null
    for (i in arr.reversedArray()) {
        val temp = Node()
        temp.data = i
        temp.next = head
        head = temp
    }

    return head
}